autogluon/utils/tabular/ml/models/xgboost/xgboost_utils.py

```diff
@@ -0,0 +1,145 @@
+    if needs_pred_proba:
+        if problem_type == MULTICLASS:
+            def function_template(y_hat, data):
+                y_true = data.get_label()
+                y_hat = _softmax(y_hat)
+                res = metric(y_true, y_hat)
+                return metric.name, -1 * res if is_higher_better else res
+        elif problem_type == SOFTCLASS:
+            def function_template(y_hat, data):
+                y_true = data.get_label()
+                y_hat = _softmax(y_hat)
+                res = metric(y_true, y_hat)
+                return metric.name, -1 * res if is_higher_better else res
+        elif problem_type == BINARY:
+            def function_template(y_hat, data):
+                y_true = data.get_label()
+                y_hat = _sigmoid(y_hat)
+                res = metric(y_true, y_hat)
+                return metric.name, -1 * res if is_higher_better else res
+        else:
+            def function_template(y_hat, data):
+                y_true = data.get_label()
+                res = metric(y_true, y_hat)
+                return metric.name, -1 * res if is_higher_better else res
+    else:
+        if problem_type == MULTICLASS:
+            def function_template(y_hat, data):
+                y_true = data.get_label()
+                y_hat = y_hat.argmax(axis=1)
+                res = metric(y_true, y_hat)
+                return metric.name, -1 * res if is_higher_better else res
+        if problem_type == SOFTCLASS:
+            def function_template(y_hat, data):
+                y_true = data.get_label()
+                y_hat = y_hat.argmax(axis=1)
+                res = metric(y_true, y_hat)
+                return metric.name, -1 * res if is_higher_better else res
+        elif problem_type == BINARY:
+            def function_template(y_hat, data):
+                y_true = data.get_label()
+                y_hat = _sigmoid(y_hat)
+                y_hat = np.round(y_hat)
+                res = metric(y_true, y_hat)
+                return metric.name, -1 * res if is_higher_better else res
+        else:
+            def function_template(y_hat, data):
+                y_true = data.get_label()
+                res = metric(y_true, y_hat)
+                return metric.name, -1 * res if is_higher_better else res
+
+    return function_template
+
+
+class OheFeatureGenerator(BaseEstimator, TransformerMixin):
+    null_category_str = '__NaN__'
+
+    def __init__(self):
+        self._feature_map = {}  # key: feature_name, value: feature_type
+        self.cat_cols = []
+        self.other_cols = []
+        self.ohe_encs = None
+        self.labels = None
+
+    def fit(self, X, y=None):
+        self.cat_cols = list(X.select_dtypes(include='category').columns)
+        self.other_cols = list(X.select_dtypes(exclude='category').columns)
+
+        self.ohe_encs = {f: OneHotEncoder(handle_unknown='ignore') for f in self.cat_cols}
```

Comment:
This can cause memory explosion / curse of dimensionality if there's a column with tons of categories.  You may consider instead using the `OneHotMergeRaresHandleUnknownEncoder` defined in: 
https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/tabular_nn/categorical_encoders.py

This should be an easy drop-in replacement and will limit the max-dimensionality of the one-hot vector by merging the rarest categories for columns with tons of unique categories.

Reply:
I will use `OneHotMergeRaresHandleUnknownEncoder` to handle categorical columns.

Reply:
great, the maximum number of categories per column should probably a hyperparameter, similar to here:
https://github.com/awslabs/autogluon/blob/master/tabular/src/autogluon/tabular/utils/ml/models/tabular_nn/hyperparameters/parameters.py#L18

Reply:
I have just set maximum number of categories to hyperparameters. Thanks for noticing me.

Reply:
awesome thanks!